src/aliceVision/sfm/pipeline/sequential/ReconstructionEngine_sequentialSfM.cpp

```diff
@@ -1415,59 +1430,285 @@ bool ReconstructionEngine_sequentialSfM::Resection(const std::size_t viewIndex)
 
+bool ReconstructionEngine_sequentialSfM::checkChieralities(
+  const Vec3& pt3D, 
+  const std::set<IndexT> & viewsId, 
+  const SfMData& scene)
+{
+  for (const IndexT & viewId : viewsId)
+  {
+    const View* view = scene.GetViews().at(viewId).get();
+    const Pose3 pose = scene.getPose(*view);
+    // Check that the point is in front of all the cameras.
+    if (pose.depth(pt3D) < 0) 
+      return false;
+  }
   return true;
 }
 
+bool ReconstructionEngine_sequentialSfM::checkAngles(const Vec3 &pt3D, const std::set<IndexT> &viewsId, const SfMData &scene, const double &kMinAngle)
+{ 
+    for (const std::size_t & viewIdA : viewsId)  
+    {
+      for (const std::size_t & viewIdB : viewsId)  
+      {
+        if (viewIdA < viewIdB)
+        {
+          double angle_deg = AngleBetweenRay(scene.getPose(*scene.GetViews().at(viewIdA).get()), 
+                                             scene.getPose(*scene.GetViews().at(viewIdB).get()),
+                                             pt3D);
+          if (angle_deg >= kMinAngle)
+            return true;
+        }      
+      }
+    }
+    return false;
+}
+
+void ReconstructionEngine_sequentialSfM::triangulateMultiViews_LORANSAC(SfMData& scene, const std::set<IndexT>& previousReconstructedViews, const std::set<IndexT>& newReconstructedViews)
+{
+  ALICEVISION_LOG_DEBUG("Triangulating (mode: multi-view LO-RANSAC)... ");
+  
+  // -- Identify the track to triangulate :
+  // This map contains all the tracks that will be triangulated (for the first time, or not)
+  // These tracks are seen by at least one new reconstructed view.
+  std::map<IndexT, std::set<IndexT>> mapTracksToTriangulate; // <trackId, observations> 
+  std::vector<IndexT> setTracksId; // <trackId>
+  {
+#pragma omp parallel for 
+    for (ptrdiff_t i = 0; i < static_cast<ptrdiff_t>(previousReconstructedViews.size()); ++i)
+    {
+      std::set<IndexT>::const_iterator iter = previousReconstructedViews.begin();
+      std::advance(iter, i);
+      const IndexT indexPrev = *iter;
+      
+      for(IndexT indexNew: newReconstructedViews)
+      {
+        if(indexPrev == indexNew)
+          continue;
+        
+        const std::size_t I = std::min((IndexT)indexNew, indexPrev);
+        const std::size_t J = std::max((IndexT)indexNew, indexPrev);
+        
+        // Find track correspondences between I and J
+        const std::set<std::size_t> set_viewIndex = { I, J };
+        track::TracksMap map_tracksCommonIJ;
+        track::TracksUtilsMap::GetTracksInImagesFast(set_viewIndex, _map_tracks, _map_tracksPerView, map_tracksCommonIJ);
+        
+        // Collect tracksIds
+        std::set<std::size_t> commonTracksId;
+        std::transform(map_tracksCommonIJ.begin(), map_tracksCommonIJ.end(),
+                       std::inserter(commonTracksId, commonTracksId.begin()),
+                       stl::RetrieveKey());
+        
+        for (std::size_t trackId : commonTracksId)
+        {
+#pragma omp critical        
```

Comment:
The whole loop is critical so better to lock directly from the start, as locking has a cost in itself.
```
for (std::size_t trackId : commonTracksId)
{
  mapTracksToTriangulate[trackId].insert(I);
  mapTracksToTriangulate[trackId].insert(J);
}
```

Reply:
Like this?

#pragma omp critical
{
for (std::size_t trackId : commonTracksId)
{
  mapTracksToTriangulate[trackId].insert(I);
  mapTracksToTriangulate[trackId].insert(J);
}
}

Reply:
yes! I missed the first line in my comment.